src/IO/Resource/ThrottlerConstraint.h

```diff
@@ -0,0 +1,200 @@
+        if (child->isActive())
+            activateChild(child.get());
+    }
+
+    void removeChild(ISchedulerNode * child_) override
+    {
+        if (child.get() == child_)
+        {
+            child_active = false; // deactivate
+            child->setParent(nullptr); // detach
+            child.reset();
+        }
+    }
+
+    ISchedulerNode * getChild(const String & child_name) override
+    {
+        if (child->basename == child_name)
+            return child.get();
+        else
+            return nullptr;
+    }
+
+    std::pair<ResourceRequest *, bool> dequeueRequest() override
+    {
+        // Dequeue request from the child
+        auto [request, child_now_active] = child->dequeueRequest();
+        if (!request)
+            return {nullptr, false};
+
+        // Request has reference to the first (closest to leaf) `constraint`, which can have `parent_constraint`.
+        // The former is initialized here dynamically and the latter is initialized once during hierarchy construction.
+        if (!request->constraint)
+            request->constraint = this;
+
+        updateBucket(request->cost);
+
+        child_active = child_now_active;
+        if (!active())
+            busy_periods++;
+        dequeued_requests++;
+        dequeued_cost += request->cost;
+        return {request, active()};
+    }
+
+    void finishRequest(ResourceRequest * request) override
+    {
+        // Recursive traverse of parent flow controls in reverse order
+        if (parent_constraint)
+            parent_constraint->finishRequest(request);
+
+        // NOTE: Token-bucket constraint does not require any action when consumption ends
+    }
+
+    void activateChild(ISchedulerNode * child_) override
+    {
+        if (child_ == child.get())
+            if (!std::exchange(child_active, true) && satisfied() && parent)
+                parent->activateChild(this);
+    }
+
+    bool isActive() override
+    {
+        return active();
+    }
+
+    size_t activeChildren() override
+    {
+        return child_active;
+    }
+
+    bool isSatisfied() override
+    {
+        return satisfied();
+    }
+
+    double getTokens() const
+    {
+        auto now = event_queue->now();
+        double elapsed = std::chrono::nanoseconds(now - last_update).count() / 1e9;
+        return std::min(tokens + max_speed * elapsed, max_burst);
+    }
+
+    std::chrono::nanoseconds getThrottlingDuration() const
+    {
+        return throttling_duration;
+    }
+
+    std::pair<double, double> getParams() const
+    {
+        return {max_burst, max_speed};
+    }
+
+private:
+    void onPostponed()
+    {
+        postponed = EventQueue::not_postponed;
+        bool was_active = active();
+        updateBucket();
+        if (!was_active && active() && parent)
+            parent->activateChild(this);
+    }
+
+    void updateBucket(ResourceCost use = 0)
+    {
+        auto now = event_queue->now();
+        if (max_speed > 0.0)
```

Comment:
Is it allowed to set `max_speed = 0` and in that case we allowed to use `max_burst` only once?